ktor-server/ktor-server-core/jvm/src/io/ktor/application/newapi/KtorPlugin.kt

```diff
@@ -0,0 +1,655 @@
+         * */
+        public fun afterResponse(vararg phases: PipelinePhase = arrayOf(ApplicationSendPipeline.After)): Unit =
+            addPhases(holder.afterResponseInterceptions, *phases)
+
+        private fun <T : Any> addPhases(target: MutableList<Interception<T>>, vararg phases: PipelinePhase) {
+            phases.forEach { phase ->
+                target.add(Interception(phase) {})
+            }
+        }
+    }
+}
+
+/**
+ * Empty implementation of [InterceptionsHolder] interface that can be used for simplicity.
+ * */
+public abstract class DefaultInterceptionsHolder(override val name: String) : InterceptionsHolder {
+    override val fallbackInterceptions: MutableList<CallInterception> = mutableListOf()
+    override val callInterceptions: MutableList<CallInterception> = mutableListOf()
+    override val monitoringInterceptions: MutableList<CallInterception> = mutableListOf()
+    override val beforeReceiveInterceptions: MutableList<ReceiveInterception> = mutableListOf()
+    override val onReceiveInterceptions: MutableList<ReceiveInterception> = mutableListOf()
+    override val beforeResponseInterceptions: MutableList<ResponseInterception> = mutableListOf()
+    override val onResponseInterceptions: MutableList<ResponseInterception> = mutableListOf()
+    override val afterResponseInterceptions: MutableList<ResponseInterception> = mutableListOf()
+}
+
+/**
+ * Gets plugin instance for this pipeline, or fails with [MissingApplicationFeatureException] if the feature is not installed
+ * @throws MissingApplicationFeatureException
+ * @param plugin plugin to lookup
+ * @return an instance of plugin
+ */
+public fun <A : Pipeline<*, ApplicationCall>> A.plugin(plugin: InterceptionsHolder): InterceptionsHolder {
+    return attributes[featureRegistryKey].getOrNull(plugin.key)
+        ?: throw MissingApplicationFeatureException(plugin.key)
+}
+
+/**
+ * [Execution] holds everything that is bound to the current plugin execution. Instance of [Execution] can be
+ * usually accessed from inside of any callback that defines feature behaviour (ex.: [KtorPlugin.onCall],
+ * [KtorPlugin.onReceive], [KtorPlugin.onResponse], etc.).
+ *
+ * [Execution] also provides useful methods, variables and mechanisms that cen be used in multiple advanced scenarios.
+ *
+ * **Please note that every method of the  [Execution] class should be considered an
+ * advanced functionality and requires a deep knowledge and understanding of the internals of Ktor.
+ * Please, read about [Pipeline] and [PipelinePhase] first.**
+ *
+ * @param SubjectT is a type of subject for the current feature that is being processed. See [subject] for more information.
+ *
+ * */
+public inline class Execution<SubjectT : Any>(private val context: PipelineContext<SubjectT, ApplicationCall>) {
+    /**
+     * Continues execution of the HTTP pipeline starting after the current plugin.
+     * Current plugin's code stops being executed from the moment of calling this function.
```

Comment:
What do you mean by `Current plugin's code stops being executed...`? I find it confusing.